src/msg/async/AsyncConnection.cc

```diff
@@ -32,6 +32,223 @@
 // Constant to limit starting sequence number to 2^31.  Nothing special about it, just a big number.  PLR
 #define SEQ_MASK  0x7fffffff
 
+void WriteQueue::fillin_iovec() {
+  /* Forward to protocol */
+  /* XXX TODO */
+}
+
+void WriteQueue::fillin_bufferlist() {
+  /* Forward to protocol */
+  /* XXX TODO */
+}
+
+bufferlist::buffers_t::const_iterator
+WriteQueue::find_buf(const bufferlist::buffers_t &bufs, unsigned int &pos) {
+  bufferlist::buffers_t::const_iterator it;
+
+  for (it = bufs.begin(); it != bufs.end() && pos; it++) {
+    auto &buf = *it;
+
+    if (pos < buf.length())
+      break;
+
+    pos -= buf.length();
+  }
+
+  return it;
+}
+
+bool WriteQueue::fillin_iovec_from_mem(void *mem,
+				       unsigned int beg,
+				       unsigned int end) {
+  struct iovec *iov = &*iovec_end_it;
+  unsigned int pos, len;
+
+  ceph_assert(MAX_SIZE > iovec_len);
+  pos = msg_end_pos - beg;
+  len = end - msg_end_pos;
+  len = min(len, MAX_SIZE - iovec_len);
+
+  iov->iov_len = len;
+  iov->iov_base = (char *)mem + pos;
```

Comment:
Since we always keep the iovec of fixed size we never know when it can be full, i.e. message can only partially fit the iovec. So when send completes we have to refill the iovec starting from buffer where we've stopped on previous iteration.  That is tricky.